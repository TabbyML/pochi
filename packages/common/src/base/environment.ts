import { Todo } from "@getpochi/tools";
import { z } from "zod";

export const Environment = z.object({
  currentTime: z.string().describe("The current time."),
  workspace: z
    .object({
      files: z.array(z.string()).describe("List of files in the workspace."),
      isTruncated: z.boolean().describe("Whether the file list is truncated."),
      activeTabs: z
        .union([
          z.array(z.string()),
          z.array(
            z.object({
              filepath: z.string().describe("The file path of the tab."),
              isActive: z
                .boolean()
                .describe("Whether this tab is currently active."),
            }),
          ),
        ])
        .optional()
        .describe("Active editor tabs opened in the current workspace."),
      activeSelection: z
        .object({
          filepath: z
            .string()
            .describe("The path of the active file selection."),
          range: z
            .object({
              start: z
                .object({
                  line: z
                    .number()
                    .describe("The starting line number of the selection."),
                  character: z
                    .number()
                    .describe(
                      "The starting character number of the selection.",
                    ),
                })
                .describe("The start position of the selection."),
              end: z
                .object({
                  line: z
                    .number()
                    .describe("The ending line number of the selection."),
                  character: z
                    .number()
                    .describe("The ending character number of the selection."),
                })
                .describe("The end position of the selection."),
            })
            .describe("The range of the active selection."),
          content: z.string().describe("The content of the active selection."),
        })
        .optional()
        .describe("Active editor selection in the current workspace."),
      gitStatus: z
        .object({
          userName: z.string().describe("The name of the git user.").optional(),
          userEmail: z
            .string()
            .describe("The email of the git user.")
            .optional(),
          origin: z
            .string()
            .optional()
            .describe("The origin URL of the git repository."),
          mainBranch: z
            .string()
            .describe("The main branch of the git repository."),
          currentBranch: z
            .string()
            .describe("The current branch of the git repository."),
          status: z.string().describe("The status of the git repository."),
          recentCommits: z
            .array(z.string())
            .describe("A list of recent git commits."),
        })
        .optional()
        .describe("Git information for the current workspace."),
      terminals: z
        .array(
          z.object({
            name: z.string().describe("The name of the terminal."),
            isActive: z.boolean().describe("Whether the terminal is active."),
            backgroundJobId: z
              .string()
              .optional()
              .describe(
                "The ID of the background job associated with the terminal.",
              ),
          }),
        )
        .optional()
        .describe("Visible terminals in the VS Code workspace."),
    })
    .describe("Information about the workspace."),
  info: z
    .object({
      cwd: z.string().describe("The current working directory."),
      shell: z.string().describe("The default shell."),
      os: z.string().describe("The operating system."),
      homedir: z.string().describe("The home directory."),
      customRules: z
        .string()
        .optional()
        .describe("Custom rules provided by the user."),
    })
    .describe("General information about the environment."),
  todos: z.array(Todo).optional().describe("Todos in current task"),
  userEdits: z
    .array(
      z.object({
        filepath: z.string().describe("Relative file path"),
        diff: z.string().describe("Diff content with inline markers"),
      }),
    )
    .optional()
    .describe("User edits since last checkpoint in the current workspace."),
});

export type Environment = z.infer<typeof Environment>;
export type GitStatus = Environment["workspace"]["gitStatus"];
