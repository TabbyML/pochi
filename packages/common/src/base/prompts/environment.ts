import type { TextUIPart, UIMessage } from "ai";
import type { Environment, GitStatus } from "../environment";
import { prompts } from "./index";

type User = { name: string; email: string };

export function createEnvironmentPrompt(
  environment: Environment,
  user: User | undefined,
) {
  const sections = [
    getSystemInfo(environment),
    getUserInfo(user),
    getWorkspaceFiles(environment.workspace, environment.info),
    getCurrentOpenedFiles(environment.workspace),
    getVisibleTerminals(environment.workspace),
    getCurrentWorkingFile(environment.workspace),
    getGitStatus(environment.workspace.gitStatus),
    getUserEdits(environment.userEdits),
    getTodos(environment.todos),
  ]
    .filter(Boolean)
    .join("\n\n");
  return sections.trim();
}

export function createLiteEnvironmentPrompt(environment: Environment) {
  const sections = [
    getCurrentOpenedFiles(environment.workspace),
    getVisibleTerminals(environment.workspace),
    getCurrentWorkingFile(environment.workspace),
    getGitStatus(environment.workspace.gitStatus),
    getUserEdits(environment.userEdits),
    getTodos(environment.todos),
  ]
    .filter(Boolean)
    .join("\n\n");
  return sections.trim();
}

function getSystemInfo(environment: Environment) {
  const { info, currentTime } = environment;
  const prompt = `# System Information

Operating System: ${info.os}
Default Shell: ${info.shell}
Home Directory: ${info.homedir}
Current Working Directory: ${info.cwd}
Current Time: ${currentTime}`;
  return prompt;
}

function getUserInfo(user: User | undefined) {
  if (!user) {
    return "";
  }

  const userInfo = [];
  if (user.name) {
    userInfo.push(`- Name: ${user.name}`);
  }

  if (user.email) {
    userInfo.push(`- Email: ${user.email}`);
  }

  if (userInfo.length > 0) {
    return `# User Information\n${userInfo.join("\n")}`;
  }

  return "";
}

function getWorkspaceFiles(
  workspace: Environment["workspace"],
  info: Environment["info"],
) {
  const { files, isTruncated } = workspace;
  const filesList = files.join("\n");
  const truncatedMessage = isTruncated
    ? "\n(Note: The list of files is truncated. Use listFiles tool to explore if needed)"
    : "";
  return `# Current Working Directory (${info.cwd}) Files\n${filesList}${truncatedMessage}`;
}

function getCurrentOpenedFiles(workspace: Environment["workspace"]) {
  const openFiles = workspace.activeTabs ?? [];
  if (openFiles.length === 0) {
    return "";
  }
  const header = `# Active File Tabs in Editor\nHere are the open file tabs in the editor. If a user mentions "this" or "that" without an active selection, they are likely referring active tab (if exists) below:`;
  return `${header}\n${openFiles
    .map((tab) => {
      if (typeof tab === "string") {
        return tab;
      }
      return tab.isActive ? `${tab.filepath} (active)` : tab.filepath;
    })
    .join("\n")}`;
}

function getVisibleTerminals(workspace: Environment["workspace"]) {
  const terminals = workspace.terminals ?? [];
  if (terminals.length === 0) {
    return "";
  }
  return `# Opened Terminals in Editor\n${terminals
    .map(
      (t) =>
        `${t.isActive ? "* " : "  "}${t.name}${t.isActive ? " (selected)" : ""}${t.backgroundJobId ? ` (background job: ${t.backgroundJobId})` : ""}`,
    )
    .join("\n")}`;
}

function getCurrentWorkingFile(workspace: Environment["workspace"]) {
  const selection = workspace.activeSelection;
  if (!selection) {
    return "";
  }
  const { filepath, range, content, notebookCell } = selection;
  if (!content || content.trim() === "") {
    return "";
  }

  const location = notebookCell
    ? `${filepath} (Cell ID: ${notebookCell.cellId})`
    : `${filepath}:${range.start.line + 1}-${range.end.line + 1}`;

  return `# Active Selection (${location})\n\n\`\`\`\n${content}\n\`\`\`\n`;
}

function getGitStatus(gitStatus: GitStatus | undefined) {
  if (!gitStatus) return "# GIT STATUS\nThis workspace is not managed by git";

  const { currentBranch, mainBranch, status, recentCommits } = gitStatus;

  let result = "# GIT STATUS\n";

  if (gitStatus.origin) {
    result += `Origin: ${gitStatus.origin}\n`;
  }
  result += `Current branch: ${currentBranch}\n`;
  result += `Main branch (you will usually use this for PRs): ${mainBranch}\n\n`;

  if (status) {
    result += `Status:\n${status}\n\n`;
  }

  if (recentCommits.length > 0) {
    result += `Recent commits:\n${recentCommits.join("\n")}`;
  }

  return result;
}

export function injectEnvironment(
  messages: UIMessage[],
  environment: Environment | undefined,
): UIMessage[] {
  if (environment === undefined) return messages;
  const messageToInject = messages.at(-1);
  if (!messageToInject) return messages;
  if (messageToInject.role !== "user") return messages;

  const { gitStatus } = environment.workspace;
  const user =
    gitStatus?.userEmail && gitStatus?.userName
      ? {
          name: gitStatus.userName,
          email: gitStatus.userEmail,
        }
      : undefined;

  const environmentDetails =
    messages.length === 1
      ? createEnvironmentPrompt(environment, user)
      : createLiteEnvironmentPrompt(environment);

  const reminderPart = {
    type: "text",
    text: prompts.createSystemReminder(environmentDetails),
  } satisfies TextUIPart;

  const parts =
    // Remove existing environment system reminders.
    messageToInject.parts.filter(
      (x) => x.type !== "text" || !prompts.isEnvironmentSystemReminder(x.text),
    ) || [];
  const lastTextPartIndex = parts.findLastIndex(
    (parts) => parts.type === "text",
  );
  // Insert remainderPart before lastTextPartIndex
  messageToInject.parts = [
    ...parts.slice(0, lastTextPartIndex),
    reminderPart,
    ...parts.slice(lastTextPartIndex),
  ];

  return messages;
}

function getTodos(todos: Environment["todos"]) {
  if (todos === undefined || todos.length === 0) {
    return "# TODOs\nNo TODOs yet, if you are working on tasks that would benefit from a todo list please use the todoWrite tool to create one.";
  }

  return `# TODOs
Here's todo list for current task. If a task is marked as cancelled or completed, it no longer needs your attention, NEVER ATTEMPT TO COMPLETE IT AGAIN, this is SUPER IMPORTANT!!!.
Otherwise, please follow the todo list to complete the task.

${JSON.stringify(todos, null, 2)}`;
}

function getUserEdits(userEdits: Environment["userEdits"]) {
  if (!userEdits || userEdits.length === 0) {
    return "";
  }

  // Format structured user edits data
  const formattedFiles = userEdits
    .map((edit) => {
      return `**${edit.filepath}** (modified)
\`\`\`diff
${edit.diff}
\`\`\``;
    })
    .join("\n\n");

  return `# User Edits
The user has made the following edits to the workspace. Please take these changes into account when proceeding with the task.

${formattedFiles}`;
}
