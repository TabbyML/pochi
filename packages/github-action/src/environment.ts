/**
 * Environment variable utilities
 */

export function readGithubToken(): string {
  const githubToken = process.env.GITHUB_TOKEN;
  if (!githubToken) {
    throw new Error(
      "GitHub token not found. Please ensure the `github-token` input is set in your workflow.",
    );
  }
  return githubToken;
}

export function readPochiConfig() {
  const pochiToken = process.env.POCHI_TOKEN;
  if (!pochiToken) {
    throw new Error(
      "POCHI_TOKEN environment variable is required. Please add your pochi session token to GitHub Secrets as POCHI_TOKEN.",
    );
  }

  return {
    token: pochiToken,
    model: process.env.POCHI_MODEL || "",
  };
}
