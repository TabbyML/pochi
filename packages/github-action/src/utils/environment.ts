/**
 * Environment variable utilities
 */
export interface EnvironmentConfig {
  model: string;
  pochiUrl: string;
  pochiToken: string;
  githubRunId: string;
  githubToken?: string;
}

export function getEnvironmentConfig(): EnvironmentConfig {
  const githubRunId = process.env.GITHUB_RUN_ID;
  if (!githubRunId) {
    throw new Error("Environment variable GITHUB_RUN_ID is not set");
  }

  const pochiToken = process.env.POCHI_SESSION_TOKEN;
  if (!pochiToken) {
    throw new Error(
      "POCHI_SESSION_TOKEN environment variable is required. Please add your pochi session token to GitHub Secrets as POCHI_SESSION_TOKEN.",
    );
  }

  return {
    model: process.env.POCHI_MODEL || "",
    pochiUrl: "https://app.getpochi.com",
    pochiToken,
    githubRunId,
    githubToken: process.env.GITHUB_TOKEN,
  };
}

export function getPochiConfig() {
  const config = getEnvironmentConfig();
  return {
    url: config.pochiUrl,
    token: config.pochiToken,
    model: config.model,
  };
}
