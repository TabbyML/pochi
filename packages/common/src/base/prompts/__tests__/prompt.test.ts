import { expect, test } from "vitest";
import { createEnvironmentPrompt } from "../environment";
import { createSystemPrompt } from "../system";

test("instructions", () => {
  expect(
    createSystemPrompt(
      `# Rules from (abc)`,
      undefined,
      "custom instructions from mcp servers",
    ),
  ).toMatchSnapshot();
});

test("snapshot", () => {
  expect(
    createSystemPrompt(`# Rules from (abc)`),
  ).toMatchSnapshot();
});

test("environment", () => {
  expect(
    createEnvironmentPrompt({
        currentTime: "2021-01-01T00:00:00.000Z",
        workspace: {
          files: ["index.ts", "package.json", "tsconfig.json", "README.md"],
          isTruncated: false,
          activeTabs: ["README.md", "tsconfig.json", "package.json"],
          activeSelection: {
            filepath: "README.md",
            range: {
              start: {
                line: 2,
                character: 0,
              },
              end: {
                line: 4,
                character: 46,
              },
            },
            content:
              "This template provides a minimal setup to get React working in Vite with HMR and some ESLint rules.\n\nCurrently, two official plugins are available:",
          },
          gitStatus: {
            origin: 'https://github.com/username/repo.git',
            currentBranch: 'add-environment-to-chat-request-body',
            mainBranch: 'main',
            status: 'M packages/vscode-webui-bridge/src/index.ts\nA packages/vscode-webui/src/lib/use-environment.ts\nM packages/vscode-webui/src/lib/vscode.ts\nM packages/vscode-webui/src/routes/chat.tsx\n?? src/fib.test.ts\n?? vitest.config.ts',
            recentCommits: [
              '02b50f727 feat(chat): add environment property to prepareRequestBody',
              '962185adb feat(webui): add new task link and pending component',
            ],
            worktree: {gitdir: '/Users/username/repo/.git/worktrees/add-environment-to-chat-request-body'},
          },
          terminals: [
            {
              name: "Terminal 1",
              isActive: true,
            },
            {
              name: "Terminal 2",
              isActive: false,
            },
            {
              name: "Terminal 3",
              isActive: false,
              backgroundJobId: "job-id-1"
            }
          ]
        },
        todos: [
          {
            content: "fix this",
            id: "1",
            status: "pending",
            priority: "high",
          },
        ],
        info: {
          cwd: "/home/user/project",
          os: "linux",
          homedir: "/home/user",
          shell: "bash",
        },
        userEdits: [{
          filepath: "path/to/file",
          diff: "-hello\n+world"
        }]}, {name: "Pochi", email: "noreply@getpochi.com"}),
  ).toMatchSnapshot();
});
