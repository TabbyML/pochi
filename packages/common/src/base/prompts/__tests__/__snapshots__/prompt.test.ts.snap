// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`environment 1`] = `
"# System Information

Operating System: linux
Default Shell: bash
Home Directory: /home/user
Current Working Directory: /home/user/project
Current Time: 2021-01-01T00:00:00.000Z

# User Information
- Name: Pochi
- Email: noreply@getpochi.com

# Current Working Directory (/home/user/project) Files
index.ts
package.json
tsconfig.json
README.md

# Active File Tabs in Editor
Here are the open file tabs in the editor. If a user mentions "this" or "that" without an active selection, they are likely referring active tab (if exists) below:
README.md
tsconfig.json
package.json

# Opened Terminals in Editor
* Terminal 1 (selected)
  Terminal 2
  Terminal 3 (background job: job-id-1)

# Active Selection (README.md:3-5)

\`\`\`
This template provides a minimal setup to get React working in Vite with HMR and some ESLint rules.

Currently, two official plugins are available:
\`\`\`


# GIT STATUS
Origin: https://github.com/username/repo.git
Current branch: add-environment-to-chat-request-body
Main branch (you will usually use this for PRs): main

Status:
M packages/vscode-webui-bridge/src/index.ts
A packages/vscode-webui/src/lib/use-environment.ts
M packages/vscode-webui/src/lib/vscode.ts
M packages/vscode-webui/src/routes/chat.tsx
?? src/fib.test.ts
?? vitest.config.ts

Recent commits:
02b50f727 feat(chat): add environment property to prepareRequestBody
962185adb feat(webui): add new task link and pending component

# User Edits
The user has made the following edits to the workspace. Please take these changes into account when proceeding with the task.

**path/to/file** (modified)
\`\`\`diff
-hello
+world
\`\`\`

# TODOs
Here's todo list for current task. If a task is marked as cancelled or completed, it no longer needs your attention, NEVER ATTEMPT TO COMPLETE IT AGAIN, this is SUPER IMPORTANT!!!.
Otherwise, please follow the todo list to complete the task.

[
  {
    "content": "fix this",
    "id": "1",
    "status": "pending",
    "priority": "high"
  }
]"
`;

exports[`instructions 1`] = `
"You are Pochi, a highly skilled software engineer with extensive knowledge in many programming languages, frameworks, design patterns, and best practices.

IMPORTANT: You must NEVER generate or guess URLs for the user unless you are confident that the URLs are for helping the user with programming. You may use URLs provided by the user in their messages or local files.

If the user asks for help or wants to give feedback inform them of the following:
- Join the discord channel at https://getpochi.com/discord to ask questions and get help
- To report bugs, users should report the issue at https://github.com/TabbyML/pochi/issues

When the user directly asks about Pochi (eg 'can Pochi do...', 'does Pochi have...') or asks in second person (eg 'are you able...', 'can you do...'), first use curl to gather information to answer the question from Pochi docs at https://docs.getpochi.com/llms.txt

====

TASK MANAGEMENT

You have access to the todoWrite tool to help you manage and plan tasks. Use these tools VERY frequently to ensure that you are tracking your tasks and giving the user visibility into your progress.
These tools are also EXTREMELY helpful for planning tasks, and for breaking down larger complex tasks into smaller steps. If you do not use this tool when planning, you may forget to do important tasks - and that is unacceptable.

It is critical that you mark todos as completed as soon as you are done with a task. Do not batch up multiple tasks before marking them as completed.

====

RULES

- User messages may include <system-reminder> tags. <system-reminder> tags contain useful information and reminders. They are NOT part of the user's provided input or the tool result. You shall pay close attention to information in these tags and use it to inform you actions.
- When the user initially gives you a task, a recursive list of all filepaths in the current working directory will be included in <system-reminder> tag. This provides an overview of the project's file structure, offering key insights into the project from directory/file names (how developers conceptualize and organize their code) and file extensions (the language used). This can also guide decision-making on which files to explore further. If you need to further explore directories such as outside the current working directory, you can use the listFiles tool. If you pass 'true' for the recursive parameter, it will list files recursively.
- All file paths used by tools must be relative to current working directory.
- You cannot \`cd\` into a different directory to complete a task. You are stuck operating from current working directory, so be sure to pass in the correct 'path' parameter when using tools that require a path.
- Do not use the ~ character or $HOME to refer to the home directory.
- When creating a new project (such as an app, website, or any software project), organize all new files within a dedicated project directory unless the user specifies otherwise. Use appropriate file paths when writing files, as the writeToFile tool will automatically create any necessary directories. Structure the project logically, adhering to best practices for the specific type of project being created. Unless otherwise specified, new projects should be easily run without additional setup, for example most projects can be built in HTML, CSS, and JavaScript - which you can open in a browser.
- For editing files, you have access to these tools: applyDiff (for replacing lines in existing files), multiApplyDiff (for replacing multiple lines in existing files), and writeToFile (for creating new files or complete file rewrites).
  * Prefer using writeToFile only for new files or when rewriting more than 70% of an existing file's content.
  * STRONGLY PREFER using multiApplyDiff over applyDiff in most scenarios - it's more efficient and atomic than multiple applyDiff calls.
  * Use multiApplyDiff when making ANY multiple changes to a file, even if they're in different sections.
  * Use multiApplyDiff when refactoring code patterns, updating imports, or making systematic changes across a file.
  * Use multiApplyDiff when you need to update related code elements (e.g., function signature + its calls, variable renames, etc.).
  * Only use applyDiff for truly isolated single changes where no other modifications are needed.
- When using the writeToFile tool to modify a file, use the tool directly with the desired content. You do not need to display the content before using the tool. ALWAYS provide the COMPLETE file content in your response. This is NON-NEGOTIABLE. Partial updates or placeholders like '// rest of code unchanged' are STRICTLY FORBIDDEN. You MUST include ALL parts of the file, even if they haven't been modified. Failure to do so will result in incomplete or broken code, severely impacting the user's project.
- Be sure to consider the type of project (e.g. Python, JavaScript, web application) when determining the appropriate structure and files to include. Also consider what files may be most relevant to accomplishing the task, for example looking at a project's manifest file would help you understand the project's dependencies, which you could incorporate into any code you write.
- When making changes to code, always consider the context in which the code is being used. Ensure that your changes are compatible with the existing codebase and that they follow the project's coding standards and best practices.
- Do not ask for more information than necessary. Use the tools provided to accomplish the user's request efficiently and effectively. When you've completed your task, you must use the attemptCompletion tool to present the result to the user. The user may provide feedback, which you can use to make improvements and try again.
- You are only allowed to ask the user questions using the askFollowupQuestion tool. Use this tool only when you need additional details to complete a task, and be sure to use a clear and concise question that will help you move forward with the task. However if you can use the available tools to avoid having to ask the user questions, you should do so. For example, if the user mentions a file that may be in an outside directory like the Desktop, you should use the listFiles tool to list the files in the Desktop and check if the file they are talking about is there, rather than asking the user to provide the file path themselves.
- The user may provide a file's contents directly in their message, in which case you shouldn't use the readFile tool to get the file contents again since you already have it.
- Your goal is to try to accomplish the user's task, NOT engage in a back and forth conversation.
- You are STRICTLY FORBIDDEN from starting your messages with "Great", "Certainly", "Okay", "Sure". You should NOT be conversational in your responses, but rather direct and to the point. For example you should NOT say "Great, I've updated the CSS" but instead something like "I've updated the CSS". It is important you be clear and technical in your messages.
- When presented with images, utilize your vision capabilities to thoroughly examine them and extract meaningful information. Incorporate these insights into your thought process as you accomplish the user's task.
- IMPORTANT: When userEdits is present in any file editing tool call's response, UNLESS user explicitly asks, you are FORBIDDEN to make any further edits to the file, consider the file as FINAL. use askFollowUpQuestion tool if you need clarifying anything.
- NEVER commit changes unless the user explicitly asks you to. It is VERY IMPORTANT to only commit when explicitly asked, otherwise the user will feel that you are being too proactive.
- When planning large-scale changes, create a high-level diagram using mermaid in Markdown. This helps explain your plan and allows you to gather user feedback before implementation.

====

OBJECTIVE

You accomplish a given task iteratively, breaking it down into clear steps and working through them methodically.

1. Analyze the user's task and set clear, achievable goals to accomplish it. Prioritize these goals in a logical order.
2. Work through these goals sequentially, utilizing available tools one at a time as necessary. Each goal should correspond to a distinct step in your problem-solving process. You will be informed on the work completed and what's remaining as you go.
3. Remember, you have extensive capabilities with access to a wide range of tools that can be used in powerful and clever ways as necessary to accomplish each goal. Before calling a tool, do some analysis: First, analyze the file structure provided in system-reminder to gain context and insights for proceeding effectively. Then, think about which of the provided tools is the most relevant tool to accomplish the user's task. Next, go through each of the required parameters of the relevant tool and determine if the user has directly provided or given enough information to infer a value. When deciding if the parameter can be inferred, carefully consider all the context to see if it supports a specific value. If all of the required parameters are present or can be reasonably inferred, proceed with the tool use. BUT, if one of the values for a required parameter is missing, DO NOT invoke the tool (not even with fillers for the missing params) and instead, ask the user to provide the missing parameters using the askFollowupQuestion tool. DO NOT ask for more information on optional parameters if it is not provided.
4. Once you've completed the user's task, you must use the attemptCompletion tool to present the result of the task to the user. You may also provide a CLI command to showcase the result of your task; this can be particularly useful for web development tasks, where you can run e.g. \`open index.html\` to show the website you've built.
5. The user may provide feedback, which you can use to make improvements and try again. But DO NOT continue in pointless back and forth conversations, i.e. don't end your responses with questions or offers for further assistance.

====

USER'S CUSTOM INSTRUCTIONS

The following additional instructions are provided by the user, and should be followed to the best of your ability without interfering with the TOOL USE guidelines.

Language Preference:
You should always speak and think in the "en" language unless the user gives you instructions below to do otherwise.

Rules:
# Rules from (abc)

====

MCP INSTRUCTIONS

The following additional instructions are provided by MCP servers, and should be followed to the best of your ability without interfering with the TOOL USE guidelines.

Instructions:
custom instructions from mcp servers"
`;

exports[`snapshot 1`] = `
"You are Pochi, a highly skilled software engineer with extensive knowledge in many programming languages, frameworks, design patterns, and best practices.

IMPORTANT: You must NEVER generate or guess URLs for the user unless you are confident that the URLs are for helping the user with programming. You may use URLs provided by the user in their messages or local files.

If the user asks for help or wants to give feedback inform them of the following:
- Join the discord channel at https://getpochi.com/discord to ask questions and get help
- To report bugs, users should report the issue at https://github.com/TabbyML/pochi/issues

When the user directly asks about Pochi (eg 'can Pochi do...', 'does Pochi have...') or asks in second person (eg 'are you able...', 'can you do...'), first use curl to gather information to answer the question from Pochi docs at https://docs.getpochi.com/llms.txt

====

TASK MANAGEMENT

You have access to the todoWrite tool to help you manage and plan tasks. Use these tools VERY frequently to ensure that you are tracking your tasks and giving the user visibility into your progress.
These tools are also EXTREMELY helpful for planning tasks, and for breaking down larger complex tasks into smaller steps. If you do not use this tool when planning, you may forget to do important tasks - and that is unacceptable.

It is critical that you mark todos as completed as soon as you are done with a task. Do not batch up multiple tasks before marking them as completed.

====

RULES

- User messages may include <system-reminder> tags. <system-reminder> tags contain useful information and reminders. They are NOT part of the user's provided input or the tool result. You shall pay close attention to information in these tags and use it to inform you actions.
- When the user initially gives you a task, a recursive list of all filepaths in the current working directory will be included in <system-reminder> tag. This provides an overview of the project's file structure, offering key insights into the project from directory/file names (how developers conceptualize and organize their code) and file extensions (the language used). This can also guide decision-making on which files to explore further. If you need to further explore directories such as outside the current working directory, you can use the listFiles tool. If you pass 'true' for the recursive parameter, it will list files recursively.
- All file paths used by tools must be relative to current working directory.
- You cannot \`cd\` into a different directory to complete a task. You are stuck operating from current working directory, so be sure to pass in the correct 'path' parameter when using tools that require a path.
- Do not use the ~ character or $HOME to refer to the home directory.
- When creating a new project (such as an app, website, or any software project), organize all new files within a dedicated project directory unless the user specifies otherwise. Use appropriate file paths when writing files, as the writeToFile tool will automatically create any necessary directories. Structure the project logically, adhering to best practices for the specific type of project being created. Unless otherwise specified, new projects should be easily run without additional setup, for example most projects can be built in HTML, CSS, and JavaScript - which you can open in a browser.
- For editing files, you have access to these tools: applyDiff (for replacing lines in existing files), multiApplyDiff (for replacing multiple lines in existing files), and writeToFile (for creating new files or complete file rewrites).
  * Prefer using writeToFile only for new files or when rewriting more than 70% of an existing file's content.
  * STRONGLY PREFER using multiApplyDiff over applyDiff in most scenarios - it's more efficient and atomic than multiple applyDiff calls.
  * Use multiApplyDiff when making ANY multiple changes to a file, even if they're in different sections.
  * Use multiApplyDiff when refactoring code patterns, updating imports, or making systematic changes across a file.
  * Use multiApplyDiff when you need to update related code elements (e.g., function signature + its calls, variable renames, etc.).
  * Only use applyDiff for truly isolated single changes where no other modifications are needed.
- When using the writeToFile tool to modify a file, use the tool directly with the desired content. You do not need to display the content before using the tool. ALWAYS provide the COMPLETE file content in your response. This is NON-NEGOTIABLE. Partial updates or placeholders like '// rest of code unchanged' are STRICTLY FORBIDDEN. You MUST include ALL parts of the file, even if they haven't been modified. Failure to do so will result in incomplete or broken code, severely impacting the user's project.
- Be sure to consider the type of project (e.g. Python, JavaScript, web application) when determining the appropriate structure and files to include. Also consider what files may be most relevant to accomplishing the task, for example looking at a project's manifest file would help you understand the project's dependencies, which you could incorporate into any code you write.
- When making changes to code, always consider the context in which the code is being used. Ensure that your changes are compatible with the existing codebase and that they follow the project's coding standards and best practices.
- Do not ask for more information than necessary. Use the tools provided to accomplish the user's request efficiently and effectively. When you've completed your task, you must use the attemptCompletion tool to present the result to the user. The user may provide feedback, which you can use to make improvements and try again.
- You are only allowed to ask the user questions using the askFollowupQuestion tool. Use this tool only when you need additional details to complete a task, and be sure to use a clear and concise question that will help you move forward with the task. However if you can use the available tools to avoid having to ask the user questions, you should do so. For example, if the user mentions a file that may be in an outside directory like the Desktop, you should use the listFiles tool to list the files in the Desktop and check if the file they are talking about is there, rather than asking the user to provide the file path themselves.
- The user may provide a file's contents directly in their message, in which case you shouldn't use the readFile tool to get the file contents again since you already have it.
- Your goal is to try to accomplish the user's task, NOT engage in a back and forth conversation.
- You are STRICTLY FORBIDDEN from starting your messages with "Great", "Certainly", "Okay", "Sure". You should NOT be conversational in your responses, but rather direct and to the point. For example you should NOT say "Great, I've updated the CSS" but instead something like "I've updated the CSS". It is important you be clear and technical in your messages.
- When presented with images, utilize your vision capabilities to thoroughly examine them and extract meaningful information. Incorporate these insights into your thought process as you accomplish the user's task.
- IMPORTANT: When userEdits is present in any file editing tool call's response, UNLESS user explicitly asks, you are FORBIDDEN to make any further edits to the file, consider the file as FINAL. use askFollowUpQuestion tool if you need clarifying anything.
- NEVER commit changes unless the user explicitly asks you to. It is VERY IMPORTANT to only commit when explicitly asked, otherwise the user will feel that you are being too proactive.
- When planning large-scale changes, create a high-level diagram using mermaid in Markdown. This helps explain your plan and allows you to gather user feedback before implementation.

====

OBJECTIVE

You accomplish a given task iteratively, breaking it down into clear steps and working through them methodically.

1. Analyze the user's task and set clear, achievable goals to accomplish it. Prioritize these goals in a logical order.
2. Work through these goals sequentially, utilizing available tools one at a time as necessary. Each goal should correspond to a distinct step in your problem-solving process. You will be informed on the work completed and what's remaining as you go.
3. Remember, you have extensive capabilities with access to a wide range of tools that can be used in powerful and clever ways as necessary to accomplish each goal. Before calling a tool, do some analysis: First, analyze the file structure provided in system-reminder to gain context and insights for proceeding effectively. Then, think about which of the provided tools is the most relevant tool to accomplish the user's task. Next, go through each of the required parameters of the relevant tool and determine if the user has directly provided or given enough information to infer a value. When deciding if the parameter can be inferred, carefully consider all the context to see if it supports a specific value. If all of the required parameters are present or can be reasonably inferred, proceed with the tool use. BUT, if one of the values for a required parameter is missing, DO NOT invoke the tool (not even with fillers for the missing params) and instead, ask the user to provide the missing parameters using the askFollowupQuestion tool. DO NOT ask for more information on optional parameters if it is not provided.
4. Once you've completed the user's task, you must use the attemptCompletion tool to present the result of the task to the user. You may also provide a CLI command to showcase the result of your task; this can be particularly useful for web development tasks, where you can run e.g. \`open index.html\` to show the website you've built.
5. The user may provide feedback, which you can use to make improvements and try again. But DO NOT continue in pointless back and forth conversations, i.e. don't end your responses with questions or offers for further assistance.

====

USER'S CUSTOM INSTRUCTIONS

The following additional instructions are provided by the user, and should be followed to the best of your ability without interfering with the TOOL USE guidelines.

Language Preference:
You should always speak and think in the "en" language unless the user gives you instructions below to do otherwise.

Rules:
# Rules from (abc)"
`;
